akka-persistence/src/test/scala/akka/persistence/journal/leveldb/JournalCompactionSpec.scala

```diff
@@ -0,0 +1,214 @@
+
+        deletionSeqNr += deletionBatchSize
+      }
+    }
+  }
+}
+
+class JournalCompactionSpec extends JournalCompactionSpecBase(SpecComponentBuilder("leveldb-JournalCompactionSpec", 500)) {
+
+  "A LevelDB-based persistent actor" must {
+    "compact the journal upon message deletions of configured persistence ids" in {
+      val watcher = TestProbe("watcher")
+      val logger = createLogger(watcher.ref)
+      val totalMessages = 4 * builder.compactionInterval // 2000
+      val deletionBatchSize = builder.compactionInterval // 500
+      var oldJournalSize = calculateJournalSize()
+
+      for (i ← 0L.until(totalMessages)) {
+        logger ! Generate
+        watcher.expectMsg(Generated(i))
+      }
+
+      var newJournalSize = calculateJournalSize()
+      assert(oldJournalSize < newJournalSize)
+
+      var deletionSeqNr = deletionBatchSize - 1
+
+      while (deletionSeqNr < totalMessages) {
+        logger ! Delete(deletionSeqNr)
+        watcher.expectMsg(DeleteMessagesSuccess(deletionSeqNr))
+
+        oldJournalSize = newJournalSize
+        newJournalSize = calculateJournalSize()
+        assert(oldJournalSize > newJournalSize)
+
+        deletionSeqNr += deletionBatchSize
+      }
+    }
+  }
+}
+
+class JournalCompactionThresholdSpec extends JournalCompactionSpecBase(SpecComponentBuilder("leveldb-JournalCompactionThresholdSpec", 500)) {
+
+  "A LevelDB-based persistent actor" must {
+    "compact the journal only after the threshold implied by the configured compaction interval has been exceeded" in {
+      val watcher = TestProbe("watcher")
+      val logger = createLogger(watcher.ref)
+      val totalMessages = 2 * builder.compactionInterval // 1000
+      val deletionBatchSize = builder.compactionInterval // 500
+      var oldJournalSize = calculateJournalSize()
+
+      for (i ← 0L.until(totalMessages)) {
+        logger ! Generate
+        watcher.expectMsg(Generated(i))
+      }
+
+      var newJournalSize = calculateJournalSize()
+      assert(oldJournalSize < newJournalSize)
+
+      var deletionSeqNr = deletionBatchSize - 2
+
+      while (deletionSeqNr < totalMessages) {
+        logger ! Delete(deletionSeqNr)
+        watcher.expectMsg(DeleteMessagesSuccess(deletionSeqNr))
+
+        oldJournalSize = newJournalSize
+        newJournalSize = calculateJournalSize()
+
+        if (deletionSeqNr < builder.compactionInterval - 1) {
+          assert(oldJournalSize <= newJournalSize)
+        } else {
+          assert(oldJournalSize > newJournalSize)
+        }
+
+        deletionSeqNr += deletionBatchSize
+      }
+    }
+  }
+}
+
+abstract class JournalCompactionSpecBase(val builder: SpecComponentBuilder) extends PersistenceSpec(builder.config) {
+
+  def createLogger(watcher: ActorRef): ActorRef = builder.createLogger(system, watcher)
+
+  def calculateJournalSize(): Long = {
+    val journal = journalDir
+
+    if (journal.exists()) {
+      FileUtils.sizeOfDirectory(journal)
+    } else {
+      0L
```

Comment:
This was actually a workaround because "calculateJournalSize" is used in the tests to calculate the size of the directory prior to its creation as well (the journal directory is created once the persistent actor is deployed). I could just initialize the journal size variable with zero, but I just wanted to use "calculateJournalSize" in a more uniform manner. This implementation would not cause any issue to the tests (e.g. false positives) since if the journal directory is not created due to any transient failure, most tests will fail anyway. However, I can obviously change this implementation and initialize the variables directly with zero, no prob.